### ISSUE TITLE
Generated Samples Do Not Resemble Digits for Unsupervised MNIST Training

### ISSUE THREAD

**Issue by siavashk** [izmailovpavel/flowgmm#9]
Thank you for sharing your code. This is an impressive work.

I am trying to train FlowGMM in an unsupervised manner (no labels) on the MNIST dataset. I am using the script under `experiments/train_flows/train_unsup.py`. I am training with default parameters.

I am currently at epoch 10 / 100 and the loss is slowly decreasing (see attached image), but the generated samples do not look like hand-written digits (see attached). Is this issue resolved when the training is closer to being finished? Or do I need to run the training script with different arguments?

![Screenshot from 2021-03-31 15-41-40](https://user-images.githubusercontent.com/1020752/113220548-37632380-9238-11eb-81a8-924ee60bfbe7.png)
![epoch_9](https://user-images.githubusercontent.com/1020752/113220623-5c579680-9238-11eb-97dd-fe4057bf1691.png)

**Comment by siavashk** [comment]
I did more literature review on this and it seems that normalizing flows learn local correlations as opposed to semantic features. For example, see this [link](https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/tutorial11/NF_image_modeling.html#Density-modeling-and-sampling).

Also from a different [paper](https://arxiv.org/pdf/2006.08545.pdf) from the same author: 

> We argue that flows are biased towards learning graphical properties of the data such as local pixel correlations (e.g. nearby pixels usually have similar colours) rather than semantic properties of the data (e.g. what objects are shown in the image).

**Comment by izmailovpavel** [comment]
Hey @siavashk! Generally, RealNVP and glow are not outstanding at modelling MNIST visually. I would expect the results to improve if you train longer though.

**Comment by siavashk** [comment]
I did let it train until the end, but the results where not much better. The multi-scale flow in the tutorial by the University of Amsterdam (link in previous comment) seems to provide slightly better results.

Do you know of a SOTA method for normalizing flows that would actually work on MNIST?

![Screenshot from 2021-04-09 17-32-46](https://user-images.githubusercontent.com/1020752/114252381-a25bdb00-9959-11eb-91f4-56e3ddc62bb8.png)

![epoch_99](https://user-images.githubusercontent.com/1020752/114252356-86583980-9959-11eb-888a-76a733e7adc2.png)

**Comment by izmailovpavel** [comment]
Hey @siavashk to be honest, not really. Flow samples look best on datasets like Celeb-A, and are surprisingly bad on MNIST. For example, here are samples from the [iResNet paper](https://arxiv.org/abs/1811.00995), better than what you got above but still not amazing:
![image](https://user-images.githubusercontent.com/14368801/115258782-4afaff00-a0ff-11eb-9c94-a37d11e9cb4a.png)

Our code gives significantly better samples if you use labels though, the supervised and semi-supervised versions compared to the unsupervised samples:
![image](https://user-images.githubusercontent.com/14368801/115259379-d70d2680-a0ff-11eb-8ca8-c022ae6262e4.png)
